apps/frontend/src/components/global/ExportCSVModal.vue

```diff
@@ -0,0 +1,324 @@
+        case fieldNames.cciIds:
+          result[fieldNames.cciIds] = (control.data.tags.cci || []).join(', ')
+          break;
+        case fieldNames.segments:
+          result[fieldNames.segments] = this.segmentsToString(control.hdf.segments)
+          break;
+        case fieldNames.waived:
+          result[fieldNames.waived] = control.hdf.waived ? 'True' : 'False'
+          break;
+        case fieldNames.waiverData:
+          result[fieldNames.waiverData] = JSON.stringify(_.get(control, 'hdf.wraps.waiver_data'))
+          break;
+      }
+    })
+    return result
+  }
+
+  convertRows(file: ProfileFile | EvaluationFile): ControlSetRows {
+    const rows: ControlSetRows = [];
+    const controls = FilteredDataModule.controls({...this.filter, fromFile: [file.unique_id]});
+    for (const ctrl of controls) {
+      rows.push(this.convertRow(file, ctrl));
+    }
+    return rows;
+  }
+
+  /**
+   * Generates the CSV for our first file so the user can change what fields will be in the final export.
+   */
+  generateCSVPreview() {
+    const file = InspecDataModule.allFiles.find(
+        (f) => f.unique_id === this.filter.fromFile[0]
+      );
+    if (file){
+      this.rows = this.convertRows(file);
+    }
+  }
+
+  truncate(string: string): string {
+    return _.truncate(string, {length: 100})
+  }
+
+  exportCSV() {
+    this.files = [];
+    // In case we have multiple files, make an array to store them
+    this.filter.fromFile.forEach(async (fileId) => {
+      let rows: ControlSetRows = [];
+      const file = InspecDataModule.allFiles.find(
+        (f) => f.unique_id === fileId
+      );
+      if(file) {
+        // Convert all controls from a file to ControlSetRows
+        rows = this.convertRows(file);
+        // Convert our rows to CSV
+        const csvString = await new ObjectsToCsv(rows).toString()
+        // If we only have one file we can save just one csv file
+        if(this.filter.fromFile.length === 1) {
+          const blob = new Blob([csvString], {
+            type: 'application/csv'
+          });
+          saveAs(blob, this.cleanUpFilename(`${file.filename}.csv`));
+        } else {
+          // Otherwise add to our files list
+          this.files.push({
+            filename: this.cleanUpFilename(`${file.filename}.csv`),
+            data: csvString
+          })
+        }
+        // If we have all of our files converted we can save them as a .zip file
+        if(this.filter.fromFile.length !== 1 && this.files.length === this.filter.fromFile.length){
+          const zipfile = new ZipFile();
+          this.files.forEach((csvFile) => {
+            const buffer = Buffer.from(csvFile.data);
+            zipfile.addBuffer(buffer, csvFile.filename);
+          })
+          zipfile.outputStream.pipe(
+            concat({encoding: 'uint8array'}, (b: Uint8Array) => {
+              saveAs(new Blob([b]), 'exported_csvs.zip');
+            })
+          );
+          zipfile.end();
```

Comment:
All of this code being inside a for loop doesn't make much sense. The only code that needs to be in the `for` loop is the code that actually populates the `files` array. Once that is populated, write the logic to check if your new array is 1 file or multiple files in length and then provide the user with an appropriate download. Also the files being global to the component is unnecessary since it is only used in this function.

Reply:
This is in a loop due to objects-to-csv only providing an async function, and javascript doesn't seem to have a callback method on forEach loops. Switched to promises now.